src/iconforge.rs

```diff
@@ -0,0 +1,1088 @@
+        Err(err) => {
+            return Err(format!("Error getting image for {}: {}", sprite_name, err));
+        }
+    })
+}
+
+/// Gives an image back to the cache, after it is done being used.
+fn return_image(image: RgbaImage, icon: &IconObject) -> Result<(), Error> {
+    zone!("insert_rgba_image");
+    if icon.icon_hash_input.is_empty() {
+        return Err(Error::IconForge(format!(
+            "No icon_hash_input generated for {}",
+            icon
+        )));
+    }
+    ICON_STATES.insert(icon.icon_hash_input.to_owned(), image);
+    Ok(())
+}
+
+fn apply_all_transforms(image: &mut RgbaImage, transforms: &Vec<Transform>) -> Result<(), String> {
+    let mut errors = Vec::<String>::new();
+    for transform in transforms {
+        if let Err(error) = transform_image(image, transform) {
+            errors.push(error);
+        }
+    }
+    if !errors.is_empty() {
+        return Err(errors.join("\n"));
+    }
+    Ok(())
+}
+
+/// Applies transforms to a RgbaImage.
+fn transform_image(image: &mut RgbaImage, transform: &Transform) -> Result<(), String> {
+    zone!("transform_image");
+    match transform {
+        Transform::BlendColor { color, blend_mode } => {
+            zone!("blend_color");
+            let mut color2: [u8; 4] = [0, 0, 0, 255];
+            {
+                zone!("from_hex");
+                let mut hex: String = color.to_owned();
+                if hex.starts_with('#') {
+                    hex = hex[1..].to_string();
+                }
+                if hex.len() == 6 {
+                    hex += "ff";
+                }
+
+                if let Err(err) = hex::decode_to_slice(hex, &mut color2) {
+                    return Err(format!("Decoding hex color {} failed: {}", color, err));
+                }
+            }
+            for x in 0..image.width() {
+                for y in 0..image.height() {
+                    let px = image.get_pixel_mut(x, y);
+                    let pixel = px.channels();
+                    let blended = Rgba::blend_u8(pixel, &color2, *blend_mode);
+
+                    *px = image::Rgba::<u8>(blended);
+                }
+            }
+        }
+        Transform::BlendIcon { icon, blend_mode } => {
+            zone!("blend_icon");
+            let (mut other_image, cached) =
+                icon_to_image(icon, &format!("Transform blend_icon {}", icon), true, false)?;
+
+            if !cached {
+                apply_all_transforms(&mut other_image, &icon.transform)?;
+            };
+            for x in 0..std::cmp::min(image.width(), other_image.width()) {
+                for y in 0..std::cmp::min(image.width(), other_image.width()) {
+                    let px1 = image.get_pixel_mut(x, y);
+                    let px2 = other_image.get_pixel(x, y);
+                    let pixel_1 = px1.channels();
+                    let pixel_2 = px2.channels();
+
+                    let blended = Rgba::blend_u8(pixel_1, pixel_2, *blend_mode);
```

Comment:
I don't *think* this way of looping is particularly amenable to performance- blend_mode is a loop invariant so this seems suitable for SIMD, but because the switch for blend_mode happens later down the line for each pixel the compiler may have a hard time taking advantage of that. Imo the switch for blend_mode should be hoisted outside of this loop, so that the actual pixel operations can be a tight loop. That is assuming the preceding body of this loop obtaining the pixel values is very cheap, that the actual blending perf even matters!

Reply:
You are correct, good catch! I can peek at the assembly and see if the compiler actually inlines the blend function. It would probably make the code much messier though, maybe I can use a macro or something.

Reply:
If you look at the tracy profiling, most of the time is spent parsing DMI files from the disk, but icon blending is still a pretty decent portion, especially for spritesheets like preferences which use a lot of operations. The main limitation is actually IO which is why everything is batched into groups and multithreaded where possible. Pretty light on actual computation.

Reply:
I don't really have a lot of time to deeply investigate the performance implications of this at the moment